### PR TITLE
Support granular "custom" MMS preferences.

### PR DESCRIPTION
**Pull request by moxie0**
Can't really test, since I'm in NZ.

### CODE REVIEW

**Comment by moxie0** [comment]
updated

File: res/values/strings.xml [hunk 1]
```diff
@@ -687,6 +692,8 @@
     <!-- arrays.xml -->
     <string name="arrays__import_export">Import / export</string>
     <string name="arrays__my_identity_key">My identity key</string>
+    <string name="arrays__use_default">Use Default</string>
```
Comment (mcginty) [inline res/values/strings.xml]: only first word should be capitalized per to design guildelines

File: res/xml/preferences_manual_mms.xml [hunk 2]
```diff
@@ -1,33 +1,36 @@
 <?xml version="1.0" encoding="utf-8"?>
 
-<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android">
-    <CheckBoxPreference android:key="pref_enable_manual_mms"
-                        android:defaultValue="false"
-                        android:title="@string/preferences__enable_manual_mms"
-                        android:summary="@string/preferences__override_system_mms_settings"/>
+<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android"
+                  xmlns:textsecure="http://schemas.android.com/apk/res-auto">
```
Comment (mcginty) [inline res/xml/preferences_manual_mms.xml]: xmlns:app to be consistent with rest of code

File: src/org/thoughtcrime/securesms/DatabaseUpgradeActivity.java [hunk 3]
```diff
@@ -58,6 +59,7 @@
   public static final int NO_V1_VERSION                        = 83;
   public static final int SIGNED_PREKEY_VERSION                = 83;
   public static final int NO_DECRYPT_QUEUE_VERSION             = 84;
+  public static final int GRANULAR_MMS_PREFERENCES             = 95;
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/DatabaseUpgradeActivity.java]: you mean 85?

File: src/org/thoughtcrime/securesms/PromptMmsActivity.java [hunk 4]
```diff
@@ -5,6 +5,8 @@
 import android.view.View;
 import android.widget.Button;
 
+import org.thoughtcrime.securesms.preferences.MmsPreferencesActivity;
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/PromptMmsActivity.java]: ?

File: src/org/thoughtcrime/securesms/database/ApnDatabase.java [hunk 5]
```diff
@@ -93,13 +94,11 @@ private ApnDatabase(final Context context) throws IOException {
                                           null,
                                           SQLiteDatabase.OPEN_READONLY | SQLiteDatabase.NO_LOCALIZED_COLLATORS);
   }
-  protected Apn getLocallyConfiguredMmsConnectionParameters() throws ApnUnavailableException {
-    if (TextSecurePreferences.isUseLocalApnsEnabled(context)) {
+
+  private Apn getCustomApnParameters() { //throws ApnUnavailableException {
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/database/ApnDatabase.java]: leftover comment

File: res/values/attrs.xml [hunk 6]
```diff
@@ -101,4 +101,5 @@
     <attr name="pref_ic_advanced" format="reference" />
 
     <attr name="app_protect_timeout_picker_color" format="reference"/>
+    <attr name="custom_pref_toggle" format="string"/>
```
Comment (mcginty) [inline res/values/attrs.xml]: Not sure what the consequences of not doing it are, but I think for view attributes you're supposed to wrap them with `<declare-styleable name="CustomDefaultPreference"><attr...></declare-styleable>`

File: src/org/thoughtcrime/securesms/database/ApnDatabase.java [hunk 7]
```diff
@@ -108,26 +107,12 @@ protected Apn getLocallyConfiguredMmsConnectionParameters() throws ApnUnavailabl
...
-      } catch (ApnUnavailableException aue) {
-        Log.w(TAG, "preference to use local apn set, but no parameters avaiable. falling back.");
-      }
-    }
-
+  public Apn getDefaultApnParameters(String mccmnc, String apn) {
     if (mccmnc == null) {
       Log.w(TAG, "mccmnc was null, returning null");
-      return null;
+      return new Apn("", "", "", "", "");
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/database/ApnDatabase.java]: should Apn have an EMPTY static instance that is this object?

File: src/org/thoughtcrime/securesms/PromptMmsActivity.java (hunk 4 above)
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PromptMmsActivity.java]: the class moved

Comment (mcginty) [inline src/org/thoughtcrime/securesms/PromptMmsActivity.java]: ahh, ok. missed it sorry.

File: res/values/attrs.xml (hunk 6 above)
Comment (moxie0) [inline res/values/attrs.xml]: is the "name" in the declare-styleable tag significant?

**Comment by moxie0** [comment]
updated

**Comment by moxie0** [comment]
updated

**Comment by mcginty** [comment]
:+1: lgtm, haven't tested

**Pull request closed without merging**